README.md
# UDP Media Router
Route UDP data between sockets in realtime

## Table of contents
- [RMRP](#rmrp)
	- [ADD](#add)
	- [REMOVE](#remove)
	- [ROUTE](#route)

## RMRP
**Realtime Media Routing Protocol**  
The router is listening for commands as text sent directly over TCP to port 8848 by default.

### ADD
`ADD <SERVER|CLIENT> <addr>`  

**Example:**  
Listen to packets on port 3001  
`ADD SERVER :3001`  

Try to connect to an open port on localhost  
`ADD CLIENT localhost:3002`

### REMOVE  
`REMOVE <addr>`  
Removes any connections for the specified address and stops forwarding/receiving data to it. Essentially the opposite of `ADD`.

**Example:**  
Remove a connection for address `localhost:3002`  
`REMOVE localhost:3002`  

Remove a connection for address `:3001`  
`REMOVE :3001`

### ROUTE
`ROUTE <source> <destination>`

**Note:**  
Both the source and destination must have been created with `ADD` before trying to route media between them.

**Example:**  
Route data from a server listening on port 3001 and a client connected to `localhost:3002`  
`ROUTE :3001 localhost:3002`

CHANGELOG.md
# Changelog

## [v1.0.0]
### Changes
- Initial version

test.js
const dgram = require('dgram')
const client = dgram.createSocket('udp4')
const server = dgram.createSocket('udp4')

let lastBuf

function wait (ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

class Client {
  constructor (host, port) {
    this.host = host
    this.port = port
    this.ready = true
  }

  sendIfReady (msg) {
    if (!this.ready) return
    this.ready = false
    client.send(msg, 0, msg.length, this.port, this.host, () => {
      this.ready = true
    })
  }
}

const CLIENTS = [
  new Client('127.0.0.1', 3002),
  new Client('127.0.0.1', 3003),
  new Client('127.0.0.1', 3004),
  new Client('127.0.0.1', 3005)
]

server.on('message', msg => {
  for (let client of CLIENTS) {
    client.sendIfReady(msg)
  }
})

server.bind(3001, '127.0.0.1')


main.go
package main

import (
	"io"
	"log"
	"github.com/axelboberg/go-rtp-repeater/internal/tcp"
	"github.com/axelboberg/go-rtp-repeater/internal/udp"
	"github.com/axelboberg/go-rtp-repeater/internal/rmrp"
)

func onListener (listener io.Reader) {
	var cerr error
	for cerr == nil {
		str, err := tcp.ReadString(listener)
		cerr = err
		
		log.Print(str)

		parts, parseErr := rmrp.Parse(str)
		
		if parseErr != nil {
			log.Println(parseErr)
			cerr = parseErr
		} else {
			switch parts[0] {
				
			// ROUTE
			case "ROUTE":
				udp.Route(parts[1], parts[2])
				
			// ADD
			case "ADD":
				if parts[1] == "SERVER" {
					go udp.CreateServer(parts[2])
				} else {
					go udp.CreateClient(parts[2])
				}
			
			// REMOVE
			case "REMOVE":
				udp.Kill(parts[1])
				
			}
		}
	}
}

func main () {	
	tcp.CreateServer(":8848", onListener)
	log.Println("Go is alive")
}

internal/udp/udp.go
package udp

import (
	"net"
	"log"
)

var conns = Connections{conns: make(map[string]*net.UDPConn)}
var routing = Routing{up: make(map[string]string), down: make(map[string]map[string]bool)}

func dataLoop (addr string) {
	for {
		conn, ok := conns.Get(addr)
		if !ok {
			break
		}
		
		buf := make([]byte, 4096)
		n, _, err := conn.ReadFromUDP(buf)
		
		if err != nil {
			log.Print(err)
			continue
		}

		// Create a slice containing
		// only the actual packet by
		// using the length returned
		// by ReadFromUDP
		conc := buf[0:n]
		
		recvs := routing.Destinations(addr)
		
		for _, recv := range recvs {
			recv_conn, recv_ok := conns.Get(recv);
			if recv_ok {
				recv_conn.Write(conc)
			}	
		}
	}
}

func CreateServer (port string) {
	if _, ok := conns.Get(port); ok {
		log.Print("[UDP] Port", port, "is already in use")
		return
	}
	
	addr, err := net.ResolveUDPAddr("udp", port)
	conn, err := net.ListenUDP("udp", addr)

	if err != nil {
		log.Print(err)
		return
	}

	conns.Add(port, conn)
	dataLoop(port)
}

func CreateClient (addr string) {
	dst, err := net.ResolveUDPAddr("udp", string(addr))
	conn, _ := net.DialUDP("udp", nil, dst)
	
	if err != nil {
		log.Println(err)
		return
	}
	
	conns.Add(addr, conn)
	dataLoop(addr)
}

func Kill (addr string) {
	if conn, ok := conns.Get(addr); ok {
		conns.Remove(addr)
		conn.Close()
	}
}

func Route (from string, to string) {
	routing.Route(from, to)
}

internal/udp/routing.go
package udp

import (
	"sync"
)

type Routing struct {
	mu sync.Mutex
	up map[string]string
	down map[string]map[string]bool
}

// Get the keys from a map
// as a slice of strings
func keys (m map[string]bool) []string {
	keys := make([]string, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	return keys
}

// Insert a new routing record safely
// by making sure that a destination
// can only have one source at a time
func (r *Routing) Route (from string, to string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	
	if _, ok := r.up[to]; ok {
		prev_from := r.up[to]
		delete(r.up, to)
		
		if _, ok := r.down[prev_from]; ok {
			delete(r.down[prev_from], to)
		}
	}
	
	if _, ok := r.down[from]; !ok {
		r.down[from] = make(map[string]bool)
	}
	
	r.down[from][to] = true
	r.up[to] = from
}

// Get all destinations for a
// source as a slice of strings
func (r *Routing) Destinations (host string) []string {
	r.mu.Lock()
	defer r.mu.Unlock()
	return keys(r.down[host])
}

internal/rmrp/rmrp.go
package rmrp

import (
	"errors"
	"strings"
)

type validator func ([]string) ([]string, error)

var verbs = map[string]validator{
	// ROUTE
	// Format:
	//	ROUTE <SOURCE> <DESTINATION>
	// Example:
	//	ROUTE :3001 localhost:3002 - Will start routing the data from a server on port 3001 
	//															 to a client connected to localhost:3002
	"ROUTE": func (parts []string) ([]string, error) {
		if len(parts) != 3 {
			return nil, errors.New("Malformed command. Should follow format \"ROUTE <input> <output>\"")
		}
		return parts, nil
	},
	
	// ADD
	// Format:
	//	ADD <SERVER|CLIENT> <addr>
	// Example:
	//	ADD SERVER :3001 					- Starts listening for data on localhost:3001
	//	ADD CLIENT localhost:3002 - Try to create a socket by connecting to localhost:3002
	"ADD": func (parts []string) ([]string, error) {
		if len(parts) != 3 {
			return nil, errors.New("Malformed command. Should follow format \"ADD <type> <addr>\"")
		}
		
		if parts[1] != "SERVER" && parts[1] != "CLIENT" {
			return nil, errors.New("Malformed command. Type must be either \"SERVER\" or \"CLIENT\"")
		}
		
		return parts, nil
	},
	
	// REMOVE
	// Format:
	//	REMOVE <addr>
	// Example:
	//	REMOVE :3001							- Removes the server or client created for the address :3001
	//	REMOVE localhost:3002			- Removes the server or client created for the address localhost:3002
	"REMOVE": func (parts []string) ([]string, error) {
		if len(parts) != 2 {
			return nil, errors.New("Malformed command. Should follow format \"REMOVE <addr>\"")
		}
		return parts, nil
	},
}

// Parse a string as a RMRP-command
func Parse (cmd string) ([]string, error) {
	clean := strings.TrimSuffix(cmd, "\n")
	parts := strings.Split(clean, " ")
	
	if validator, ok := verbs[parts[0]]; ok {
		return validator(parts)
	}
	
	err := errors.New("Invalid command")
	return nil, err
}